### PR TITLE
GB10: add Bonsai 2 27B benchmarks with the DSpark v2 drafter

### PR DESCRIPTION
**Pull request by usmaneth**
## Summary

Supersedes #168.

Adds the DGX Spark (GB10) entry for Ternary-Bonsai-2-27B with speculative decoding, plus one index row in each README.

- llama-bench: PQ2_0 919.05 t/s pp512 / 27.79 t/s tg128; PTQ1_0 432.59 / 32.83 (build 5d80cff0b). Plain decode is flat from 64 to 2,000 generated tokens (29.7-29.9 t/s on an idle GPU).
- llama-server workload matrix, 62 prompts (code, math, reasoning, chat, long-form, tool calls, agent turns, six 2,000-token prompts, the quicksort prompt of the older entry), temperature 0, seed 42, exact-match verification, idle GPU:
  - DSpark v2 (`Ternary-Bonsai-2-27B-dspark-dflash-v2-Q4_K_M.gguf`, trained for this target, K=5): 57.4 t/s blended over 40 prompts (1.93x), 74.7 on math (2.52x), 65.7 on code (2.22x), 66.8 on the 2,000-token set (2.26x).
  - DSpark v1 (the Ternary-Bonsai-27B drafter re-converted for this target, K=4): 57.7 blended (1.94x); leads on reasoning, tool calls and agent turns.
  - Multi-slot (1/2/4 slots), power (mJ per generated token: 2,135 without a drafter, 1,495 with DSpark v2), tool-call validity (6/8 on every configuration, baseline included) and output identity (37 of 65 outputs identical to the no-drafter run for both drafters, with the same 28 divergence points, which come from the batched verify pass and not from the draft).
- Both drafters need a runtime fix: on prism 1a07bfa5f and on the release tag that setup.sh installs, any drafter that borrows the target's token embedding and output head accepts 0.4-2.0% of drafted tokens, with no warning. The entry cites the fix (branch `fix/dflash-borrowed-hadamard`, PR title "dflash: apply the target's Hadamard transforms to borrowed embeddings and head"), gives the clean-vs-fixed acceptance table, and gives the reproduction commands.
- Launcher notes: the `*dspark-dflash*.gguf` glob picks the Q4_0 file first when both drafters are present, and `common.sh` reads `dspark.dspark.block_size`, which converted files do not carry, so `BONSAI_SPEC_NMAX=5` is needed for DSpark v2.

## Method

Same table shapes as `cuda-gb10-27b-linux.md`: llama-bench tables, a 3-pass single-prompt table, a workload matrix, then Multi-slot, Power, Exactness, training notes, Configuration, Reproduction and Hardware. The numbers come from `scripts/spec_bench` (separate PR); the download location of the DSpark v2 file is a placeholder until the upload.

## Validation

- `sh -n scripts/common.sh scripts/download_models.sh setup.sh`: OK
- `python3 -c "import ast; ast.parse(open('scripts/mlx_generate.py').read())"`: OK
- `git diff --check`: clean; index rows sorted by tg128; all relative links resolve; every table has consistent column counts
- Non-ASCII in the new file is limited to the H1 dashes and the `±` of the raw llama-bench rows, as in the sibling entries

### CODE REVIEW

File: community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+- NVIDIA driver 580.178.04; CUDA toolkit 13.0.88
+- Two DGX Spark nodes with the same image; llama-bench ran on node 1, the DSpark runs on node 2 with an idle GPU. Runs that overlapped other GPU work were discarded.
+
+## Reproduction
+
+Run these commands from the demo checkout after `BONSAI_FAMILY=bonsai2 BONSAI_MODEL=27B ./setup.sh`. The setup installs release binaries without the fix, so step 2 builds the runtime from source. The download location of the DSpark v2 file is not published yet; replace `<HF-REPO>` when it is.
+
+```bash
+# 1. drafter (1.03 GiB); keep it as the only *dspark-dflash*.gguf in the directory (see the launcher notes)
+hf download <HF-REPO> Ternary-Bonsai-2-27B-dspark-dflash-v2-Q4_K_M.gguf --local-dir models/bonsai2-gguf/27B
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md:222]: Agreed - this waits on the artifact upload. The DSpark v2 file is packaged (sha256 in the entry's Configuration section will be added with the link); the placeholder is replaced with the repository and revision before this is merged, and the entry is not indexed as reproducible until then.

File: community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+# 2. runtime: prism 1a07bfa5f plus the fix, CUDA build for the GB10
+git clone https://github.com/PrismML-Eng/llama.cpp.git
+git -C llama.cpp checkout 1a07bfa5f
+git -C llama.cpp apply dflash-borrowed-hadamard.patch   # the patch of the fix PR; check out its merge commit instead once it exists
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md]: Agreed. The fix branch is committed locally and goes up as a PR against PrismML-Eng/llama.cpp; once that PR exists this step changes to `curl -L <pr>.patch | git apply` (or the merge commit) and the file name goes away. Left open until the PR link is in.

File: community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+# DSpark v1: the Ternary-Bonsai-27B drafter, re-converted for this target (6 draft layers, block size 4)
+-md models/bonsai2-gguf/27B/Ternary-Bonsai-2-27B-dspark-dflash-Q4_0.gguf --spec-type draft-dspark --spec-draft-n-max 4 -ngld 999
+```
+
+PrismML ships no drafter for Bonsai 2: `scripts/download_models.sh` fetches none for this family, and the Bonsai 2 GGUF repositories hold none. The DSpark v1 file is the Ternary-Bonsai-27B drafter (`Ternary-Bonsai-27B-dspark-bf16.gguf`, trained for the older model), re-converted against the Bonsai 2 `PQ2_0` target with `gguf_dspark_to_dflash.py --drop-shared-tensors` and quantized to `Q4_0`. It borrows the token embedding and the output head from the Bonsai 2 target. With the runtime fix, it transfers to this target. Its GGUF header keeps `general.name = Bonsai-27B-dspark`, `dflash.target_layers = [2, 17, 32, 47, 62]` and block size 4. Its rows show a cross-generation drafter for comparison, and the tables label it "DSpark v1 (Ternary-Bonsai-27B drafter, K=4)". `--spec-draft-n-max 5` clamps to its block size 4 with a warning, so the n-max 4 numbers stand for both settings.
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md:60]: Fixed in 2190ba1. SPECULATIVE.md now scopes the paired-drafter promise to Bonsai-27B and Ternary-Bonsai-27B, and a new paragraph states that Bonsai 2 ships no drafter, that the downloader fetches none for it, and points to this entry for a drafter that works with the runtime fix.

File: community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+
+Notes for `BONSAI_SPECULATIVE=1 ./scripts/start_llama_server.sh`:
+
+- The launcher takes the first `*dspark-dflash*.gguf` in the model directory in glob order. `...-dspark-dflash-Q4_0.gguf` sorts before `...-dspark-dflash-v2-Q4_K_M.gguf`, so with both files present the launcher picks DSpark v1. Keep one drafter in the directory, or pass `-md` directly.
+- The launcher reads the draft length from the GGUF key `dspark.dspark.block_size` (`bonsai_dspark_block_size` in `scripts/common.sh`). A converted file carries `dflash.block_size` instead, so the launcher falls back to n-max 4. Set `BONSAI_SPEC_NMAX=5` for DSpark v2. The runtime accepts any n-max up to the block size and clamps larger values; n-max 5 gave the best rate for DSpark v2 in our sweeps.
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md]: Fixed in 2190ba1. The bullet now states the measurement: on prism 1a07bfa5f plus the fix, DSpark v2 (block size 7) ran at n-max 5 through the whole benchmark with 5 drafted tokens per round and no crash, and an n-max above the block size is clamped with a warning. SPECULATIVE.md notes that the equal-to-block-size rule reflects an earlier runtime and cites this measurement.

File: community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
...Result: 37 of 65 outputs are identical for DSpark v1, and 37 of 65 for DSpark v2. The 28 outputs that differ are the same 28 for both drafters. The first difference is at the same 0-based position for both. Examples: reasoning-06 at token 1, long-form-03 at token 9, code-06 at token 59, math-01 at token 255 and tool-05 at character 1773. Two different drafters do not produce the same 28 divergence points if the draft causes them. The position depends on the prompt, not on the draft. The cause is the verify pass. It scores several positions in one batch, and the batched kernels round differently from single-row decode. At a near-tie token the argmax flips. Given the batched logits, acceptance is exact. The no-drafter server gives identical output across repeated passes. So this entry does not claim byte-identical output. Speculative decoding on this runtime is exact given the batched logits, and identical to plain greedy decoding on 37 of 65 outputs here. The 200-token probes in Configuration reproduce count for count and byte for byte across the two patched builds. That is a different claim: the same fix on two binaries on the same code path, not drafter output against a plain run.
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md:185]: Fixed in 2190ba1. SPECULATIVE.md now says acceptance is exact against the batched verify logits, that the batched verify pass can round differently from single-row decode at near-tie tokens, so byte-identical output is not guaranteed, and cites this entry's 37 of 65 identical outputs for both drafters.

File: community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+}' | python3 -c 'import sys, json; t = json.load(sys.stdin)["timings"]; print("accepted", t["draft_n_accepted"], "of", t["draft_n"], "drafted;", round(t["predicted_per_second"], 1), "tok/s")'
+```
+
+Acceptance for a server run is `timings.draft_n_accepted / timings.draft_n` from the `/completion` response; `/v1/chat/completions` carries the same object. The server puts these counters in the response only when a drafter is loaded. Through the chat template (`--jinja`) the model thinks first, so a short `max_tokens` ends inside the reasoning block with `finish_reason: length`.
+
+To run through the demo launcher instead of step 4, copy the patched binaries over the release ones and set the draft length by hand:
+
+```bash
+cp llama.cpp/build-cuda/bin/llama-server llama.cpp/build-cuda/bin/llama-speculative-simple \
+   llama.cpp/build-cuda/bin/llama-bench llama.cpp/build-cuda/bin/*.so* bin/cuda/
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md:264]: Fixed in 2190ba1. The launcher block now runs mkdir -p bin/cuda before the copy, since the setup installs CPU binaries on aarch64.

File: community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (usmaneth) [reply community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md]: Done. The fix is public as PrismML-Eng/llama.cpp#210 from `usmaneth/llama.cpp`, commit `288859a96` (prism `1a07bfa5f` plus the fix). The build step now fetches that branch and checks out the pinned commit (5419eaf); the Configuration section links the pull request. The merge commit replaces the pin when the PR lands.

File: community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (usmaneth) [reply community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md:222]: Update: the runtime side is public now (PrismML-Eng/llama.cpp#210, pinned in 5419eaf). The artifact upload is still pending. The placeholder stays, and the entry stays out of the index, until the repository and revision exist.

File: community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+Both drafters need a runtime fix. No released binary runs any drafter on this target without it: the released runtime accepts 0.4-2.0% of drafted tokens and gives no warning. See Configuration.
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md:9]: Acknowledged. Clarified the drafter scope in the benchmark documentation.

File: community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+
+- The launcher takes the first `*dspark-dflash*.gguf` in the model directory in glob order. `...-dspark-dflash-Q4_0.gguf` sorts before `...-dspark-dflash-v2-Q4_K_M.gguf`, so with both files present the launcher picks DSpark v1. Keep one drafter in the directory, or pass `-md` directly.
+- The launcher reads the draft length from the GGUF key `dspark.dspark.block_size` (`bonsai_dspark_block_size` in `scripts/common.sh`). A converted file carries `dflash.block_size` instead, so the launcher falls back to n-max 4. Set `BONSAI_SPEC_NMAX=5` for DSpark v2 (the central environment reference still lists this variable as PowerShell-only; the Linux launcher reads it too). `scripts/common.sh` and [SPECULATIVE.md](../../SPECULATIVE.md) state that n-max must equal the block size and that a smaller value crashes; that rule reflects an earlier runtime. On prism `1a07bfa5f` plus the fix, DSpark v2 (block size 7) ran at n-max 5 through the whole benchmark. Every draft round drafted 5 tokens, and no run crashed. An n-max above the block size is clamped to the block size with a warning. n-max 5 gave the best rate for DSpark v2 in our sweeps.
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply community-benchmarks/ternary-bonsai/cuda-gb10-bonsai2-27b-linux.md:65]: Acknowledged. Updated environment variables reference to document Linux launcher support.

**Comment by ADA-quart** [comment]
Thanks for the report and the pointers in SPECULATIVE.md — I reproduced the DSpark v1 route on a
16 GB **Volta** card (V100-SXM2, Windows/CUDA, native sm_70) to see whether the fix travels down to
older compute capabilities. Reproducer and raw numbers are below; happy to turn this into a
community-benchmark row if you think it belongs in the same table shape as the GB10 entry.

**Build/drafter**: PrismML-Eng/llama.cpp `288859a96` (prism `1a07bfa5f` + PR #210, build 10707,
sm_70), drafter = `Ternary-Bonsai-27B-dspark-bf16.gguf` → `gguf_dspark_to_dflash.py
--drop-shared-tensors` (donor = Bonsai 2 PQ2_0) → `llama-quantize ... Q4_0` (588 MB,
`dflash.block_size = 4`, borrowed embedding/head). Server: `-ngl 999 -fa on -c 16384 -np 1
--jinja --temp 0`; drafter arm adds `-md ... --spec-type draft-dspark --spec-draft-n-max 4
-ngld 999`. One warm-up request per arm, `cache_prompt: false`, max_tokens 300.

| task | no drafter | DSpark v1 | speedup | draft_n | accepted | accept% |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| math | 45.4 t/s | 61.0-62.5 t/s | 1.34-1.38x | 350 | 211 | 60.3% |
| code (CSV parser) | 45.4 t/s | 41.4-41.5 t/s | **0.91x** | 549 | 161 | 29.3% |
| code (binary search) | 45.3 t/s | 50.3-50.4 t/s | 1.11x | 441 | 188 | 42.6% |
| chat | 45.6 t/s | 50.1-50.2 t/s | 1.10x | 272 | 112 | 41.2% |
| blended | 45.4 t/s | 50.9 t/s | 1.12x | - | - | 40.9% |

Two observations that may be useful:

1. **The PR #210 fix is doing exactly what the entry says on Volta.** Acceptance is 29-60% here
   against the 0.4-2.0% the clean runtime gives for a borrowed-embedding drafter on this target, and
   the run is deterministic at temperature 0 (identical `draft_n`/`accepted` across two identical
   runs).
2. **The payoff scales down hard with fewer FLOPs/watt.** Blended 1.12x on the V100 vs your 1.94x on
   GB10; the CSV-parser prompt is a *net slowdown* at 29% acceptance. The verify pass plus the
   drafter's own forward pass costs relatively more when baseline decode is already ~22 ms/token.
   Worth calling out in the entry that sub-35% acceptance is a loss on this class of card, so people
   do not enable it blindly.

Also cross-posting a one-liner on #212 since the maintainers asked for V100 acceptance/throughput
there. Thanks again for SPECULATIVE.md — the `BONSAI_SPEC_NMAX`/block-size note is what made the
conversion load correctly the first time.

**Comment by usmaneth** [comment]
@bri-prism @khosravipasha the inline review threads are resolved. Could you take a fresh look at this PR?
